tenders/spt/spt_launch_ppc64le.S

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2019 Contributors as noted in the AUTHORS file
+ *
+ * This file is part of Solo5, a sandboxed execution environment.
+ *
+ * Permission to use, copy, modify, and/or distribute this software
+ * for any purpose with or without fee is hereby granted, provided
+ * that the above copyright notice and this permission notice appear
+ * in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
+ * WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
+ * AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR
+ * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS
+ * OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
+ * NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+ * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#define ENTRY(x) .text; .globl x; .type x,%function; x:
+#define END(x)   .size x, . - x
+
+/* TODO/XXX: This calling convention is probably not 100% correct */
+
+/*
+ * r3: stack pointer to use (0x1ffffff0)
+ * r4: function to call     (0x102dd0)
+ * r5: argument to pass to function (0x11000)
+ */
+ENTRY(spt_launch)
+	mr %r1, %r3       // load stack pointer
+	stdu %r1,-16(%r1) // adjust stack pointer a little more...
```

Comment:
If I understand this correctly, it is subtracting another `0x10` from the stack pointer. What is the purpose of this, exactly? The comment is not very helpful. The reason I'm asking is that `sp` is initialised in `spt_run` by the caller to `spt->mem_size - 0x10`, so I'm wondering why not just change that to `spt->mem_size - 0x20`. Presumably this has to do with ABI stack alignment requirements, can you point me to those?

Also, can you drop the actual values from the comment heading up this function?

Reply:
This line is due to this code here:

```
0000000000000000 <_start>:
   0:   00 00 4c 3c     addis   r2,r12,0
   4:   00 00 42 38     addi    r2,r2,0
   8:   a6 02 08 7c     mflr    r0        # move from link register to r0
   c:   f0 ff c1 fb     std     r30,-16(r1)
  10:   f8 ff e1 fb     std     r31,-8(r1)
  14:   78 1b 7e 7c     mr      r30,r3
  18:   10 00 01 f8     std     r0,16(r1)   <---- write r0 to 0x10(r1) --> bang!
```

r1 is the stack pointer. What it seems to be doing is writing the address in r0 to 0x10(r1), which is beyond the allocated memory, so it blows up. It doesn't seem to be related to stack alignment, which is 0x10 (2**4) following the objdump here.

```
$ objdump -p tests/test_hello/test_hello.spt

tests/test_hello/test_hello.spt:     file format elf64-powerpcle

Program Header:
    LOAD off    0x0000000000010000 vaddr 0x0000000000100000 paddr 0x0000000000100000 align 2**16
         filesz 0x00000000000103f0 memsz 0x00000000000103f0 flags r-x
    LOAD off    0x0000000000030000 vaddr 0x0000000000120000 paddr 0x0000000000120000 align 2**16
         filesz 0x0000000000000018 memsz 0x0000000000010050 flags rw-
   STACK off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**4
         filesz 0x0000000000000000 memsz 0x0000000000000000 flags rw-
private flags = 0x2: [abiv2]
```

Reply:
Here's a link that may give some insights: https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi-1.9.html#STACK

_start happens to only access 0x10(r1) to save the link register. Though the accesses seem to be able to extend to a lot more if floating point registers were to have to be saved...

Reply:
Here's a link with the stack layout: https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi-1.9.html#STACK

An artifact of the _start function is that it 'only' accesses 0x10(r1). On the above page they talk about